include/range/v3/action/shuffle.hpp

```diff
@@ -0,0 +1,108 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Filip Matzner 2015
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_ACTION_SHUFFLE_HPP
+#define RANGES_V3_ACTION_SHUFFLE_HPP
+
+#include <functional>
+#include <range/v3/range_fwd.hpp>
+#include <range/v3/action/action.hpp>
+#include <range/v3/algorithm/shuffle.hpp>
+#include <range/v3/utility/functional.hpp>
+#include <range/v3/utility/iterator_concepts.hpp>
+#include <range/v3/utility/iterator_traits.hpp>
+#include <range/v3/utility/static_const.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        /// \addtogroup group-actions
+        /// @{
+        namespace action
+        {
+            struct shuffle_fn
+            {
+            private:
+                friend action_access;
+                template<typename Gen>
+                static auto bind(shuffle_fn shuffle, Gen && gen)
+                RANGES_DECLTYPE_AUTO_RETURN
```

Comment:
This `bind` function should be constrained with `UniformRandomNumberGenerator<Gen>`.